File: languages/erlang/web3_client.erl
-module(web3_client).
-behaviour(gen_server).

%% API
-export([start_link/0, get_block_number/0, get_balance/1, get_gas_price/0]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).

-define(RPC_URL, "https://mainnet.infura.io/v3/YOUR_KEY").

%%%===================================================================
%%% API
%%%===================================================================

start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

get_block_number() ->
    gen_server:call(?MODULE, get_block_number).

get_balance(Address) ->
    gen_server:call(?MODULE, {get_balance, Address}).

get_gas_price() ->
    gen_server:call(?MODULE, get_gas_price).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

init([]) ->
    {ok, #{rpc_url => ?RPC_URL}}.

handle_call(get_block_number, _From, State) ->
    Result = rpc_call("eth_blockNumber", []),
    {reply, Result, State};

handle_call({get_balance, Address}, _From, State) ->
    Result = rpc_call("eth_getBalance", [Address, <<"latest">>]),
    {reply, Result, State};

handle_call(get_gas_price, _From, State) ->
    Result = rpc_call("eth_gasPrice", []),
    {reply, Result, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

%%%===================================================================
%%% Internal functions
%%%===================================================================

rpc_call(Method, Params) ->
    Body = jsx:encode(#{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => 1,
        <<"method">> => list_to_binary(Method),
        <<"params">> => Params
    }),

    Headers = [{<<"Content-Type">>, <<"application/json">>}],

    case httpc:request(post, {?RPC_URL, Headers, "application/json", Body}, [], []) of
        {ok, {{_, 200, _}, _, ResponseBody}} ->
            case jsx:decode(list_to_binary(ResponseBody), [return_maps]) of
                #{<<"result">> := Result} -> {ok, Result};
                #{<<"error">> := Error} -> {error, Error};
                _ -> {error, invalid_response}
            end;
        {error, Reason} ->
            {error, Reason}
    end.

File: languages/elixir/web3_client.ex
defmodule Web3Client do
  @moduledoc """
  Ethereum Web3 client for Elixir with GenServer support
  """
  use GenServer
  require Logger

  @rpc_url "https://mainnet.infura.io/v3/YOUR_KEY"

  # Client API
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  def get_block_number do
    GenServer.call(__MODULE__, :get_block_number)
  end

  def get_balance(address) do
    GenServer.call(__MODULE__, {:get_balance, address})
  end

  def get_gas_price do
    GenServer.call(__MODULE__, :get_gas_price)
  end

  # Server Callbacks
  @impl true
  def init(_opts) do
    {:ok, %{rpc_url: @rpc_url}}
  end

  @impl true
  def handle_call(:get_block_number, _from, state) do
    result = rpc_call("eth_blockNumber", [])
    {:reply, result, state}
  end

  @impl true
  def handle_call({:get_balance, address}, _from, state) do
    result = rpc_call("eth_getBalance", [address, "latest"])
    {:reply, result, state}
  end

  @impl true
  def handle_call(:get_gas_price, _from, state) do
    result = rpc_call("eth_gasPrice", [])
    {:reply, result, state}
  end

  # Private Functions
  defp rpc_call(method, params) do
    body = Jason.encode!(%{
      jsonrpc: "2.0",
      id: 1,
      method: method,
      params: params
    })

    headers = [{"Content-Type", "application/json"}]

    case HTTPoison.post(@rpc_url, body, headers) do
      {:ok, %{body: response_body}} ->
        case Jason.decode(response_body) do
          {:ok, %{"result" => result}} -> {:ok, result}
          {:ok, %{"error" => error}} -> {:error, error}
          _ -> {:error, :invalid_response}
        end
      {:error, reason} -> {:error, reason}
    end
  end
end

defmodule WalletBalance do
  def check_multiple(addresses) do
    addresses
    |> Task.async_stream(fn address ->
      {:ok, balance} = Web3Client.get_balance(address)
      {address, hex_to_decimal(balance)}
    end)
    |> Enum.map(fn {:ok, result} -> result end)
    |> Map.new()
  end

  defp hex_to_decimal("0x" <> hex) do
    String.to_integer(hex, 16)
  end
end
